### PR TITLE
bug: migrating minio object storage (longhorn to rook migration).

### PR DESCRIPTION
**Pull request by ricardomaraschini**
#### What this PR does / why we need it:
When migrating to Rook in a cluster where Minio is running we need to migrate all object storage from the latter. This commit creates the function to migrate the data. This function will called from the Rook add-on installer (coming up in a next PR).

Also with this commit:

- Removes default storage class annotation once the data has been migrated.
- Only attempts to scale down and up Prometheus if it is installed.

### CODE REVIEW

File: scripts/common/object_store.sh [hunk 1]
```diff
@@ -257,3 +257,146 @@ EOF
     printf "\n${GREEN}Object store migration completed successfully${NC}\n"
     report_addon_success "rook-ceph-to-minio" "v1"
 }
+
+function migrate_minio_to_rgw() {
```
Comment (ricardomaraschini) [inline scripts/common/object_store.sh:261]: This is an almost equal copy of `migrate_rgw_to_minio` above.

Comment (laverya) [reply scripts/common/object_store.sh:261]: I hate to ask you to rewrite this
would it be possible to make both of these into wrapper scripts that set things like "source/dest ns" "source/dest service" "source/dest access key ID" and then call the same underlying logic?

File: scripts/common/object_store.sh [hunk 2]
```diff
@@ -257,3 +257,146 @@ EOF
...
+        if [ "$s3_url" = "http://${MINIO_HOST}" ]; then
+            kubectl -n velero patch backupstoragelocation default --type=merge -p "{\"spec\":{\"config\":{\"s3Url\":\"http://${RGW_HOST}\",\"publicUrl\":\"http://${rgw_ip}\"}}}"
+
+            while read -r resticrepo; do
+                oldResticIdentifier=$(kubectl -n velero get resticrepositories "$resticrepo" -ojsonpath="{ .spec.resticIdentifier }")
+                newResticIdentifier=$(echo "$oldResticIdentifier" | sed "s/${MINIO_HOST}/${RGW_HOST}/")
+                kubectl -n velero patch resticrepositories "$resticrepo" --type=merge -p "{\"spec\":{\"resticIdentifier\":\"${newResticIdentifier}\"}}"
+            done < <(kubectl -n velero get resticrepositories --selector=velero.io/storage-location=default --no-headers | awk '{ print $1 }')
+        else
+            echo "default backupstoragelocation was not minio, skipping"
```
Comment (emosbaugh) [inline scripts/common/object_store.sh:377]:
```suggestion
            echo "default velero backupstoragelocation is not minio, skipping"
```

File: scripts/common/object_store.sh [hunk 3]
```diff
@@ -257,3 +257,146 @@ EOF
...
+    fi
+
+    local rgw_ip=$($DIR/bin/kurl format-address "$RGW_CLUSTER_IP")
+    # Update registry to use rook
+    if kubernetes_resource_exists kurl configmap registry-config; then
+        echo "Updating registry to use rook"
+        kubectl -n kurl get configmap registry-config -ojsonpath='{ .data.config\.yml }' | sed "s/regionendpoint: http.*/regionendpoint: http:\/\/${rgw_ip}/" > config.yml
+        kubectl -n kurl delete configmap registry-config
+        kubectl -n kurl create configmap registry-config --from-file=config.yml=config.yml
+        rm config.yml
```
Comment (emosbaugh) [inline scripts/common/object_store.sh:354]:
```suggestion
        echo "Updating registry to use rook"
        local temp_file=
        temp_file=$(mktemp --suffix=.yml)
        kubectl -n kurl get configmap registry-config -ojsonpath='{ .data.config\.yml }' | sed "s/regionendpoint: http.*/regionendpoint: http:\/\/${rgw_ip}/" > "$temp_file"
        kubectl -n kurl delete configmap registry-config
        kubectl -n kurl create configmap registry-config --from-file=config.yml="$temp_file"
        rm "$temp_file"
```

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: scripts/common/object_store.sh [hunk 4]
```diff
@@ -202,49 +198,49 @@ EOF
...
-            sed -i "s/aws_access_key_id=.*/aws_access_key_id=${MINIO_ACCESS_KEY_ID}/" cloud
-            sed -i "s/aws_secret_access_key=.*/aws_secret_access_key=${MINIO_ACCESS_KEY_SECRET}/" cloud
+            sed -i "s/aws_access_key_id=.*/aws_access_key_id=${destination_access_key}/" cloud
+            sed -i "s/aws_secret_access_key=.*/aws_secret_access_key=${destination_secret_key}/" cloud
             cloud=$(cat cloud | base64 -w 0)
             kubectl -n velero patch secret cloud-credentials -p "{\"data\":{\"cloud\":\"${cloud}\"}}"
             rm cloud
         else
-            echo "cloud-credentials secret were not for rgw, skipping"
+            echo "cloud-credentials secret were not for the source host, skipping"
```
Comment (emosbaugh) [inline scripts/common/object_store.sh:243]: something is wrong about this sentence

File: scripts/common/object_store.sh [hunk 5]
```diff
@@ -157,20 +154,19 @@ spec:
...
-    - --source_access_key_secret=$RGW_ACCESS_KEY_SECRET
-    - --dest_host=$MINIO_HOST
-    - --dest_access_key_id=$MINIO_ACCESS_KEY_ID
-    - --dest_access_key_secret=$MINIO_ACCESS_KEY_SECRET
+    - --source_host=$source_host
+    - --source_access_key_id=$source_access_key
+    - --source_access_key_secret=$source_secret_key
+    - --dest_host=$destination_host
+    - --dest_access_key_id=$destination_access_key
+    - --dest_access_key_secret=$destination_secret_key
```
Comment (ricardomaraschini) [inline scripts/common/object_store.sh:164]: I think we already have a function that copies object stores [here](https://github.com/replicatedhq/kURL/blob/main/scripts/common/object_store.sh#L69-L72). Maybe we can replace this with a call to `migrate_object_store`.

Comment (laverya) [reply scripts/common/object_store.sh:164]: They're not quiiiiiite the same, but I'll give it a go

Comment (laverya) [reply scripts/common/object_store.sh]: easy enough

File: scripts/common/object_store.sh [hunk 6]
```diff
@@ -204,49 +167,53 @@ EOF
...
-            kubectl -n velero get secret cloud-credentials -ojsonpath='{ .data.cloud }' | base64 -d > cloud
-            sed -i "s/aws_access_key_id=.*/aws_access_key_id=${MINIO_ACCESS_KEY_ID}/" cloud
-            sed -i "s/aws_secret_access_key=.*/aws_secret_access_key=${MINIO_ACCESS_KEY_SECRET}/" cloud
-            cloud=$(cat cloud | base64 -w 0)
+        if kubectl -n velero get secret cloud-credentials -ojsonpath='{ .data.cloud }' | base64 -d | grep -q "$source_access_key"; then
+            local temp_file=
+            temp_file=$(mktemp)
+            kubectl -n velero get secret cloud-credentials -ojsonpath='{ .data.cloud }' | base64 -d > "$temp_file"
+            sed -i "s/aws_access_key_id=.*/aws_access_key_id=${destination_access_key}/" cloud
+            sed -i "s/aws_secret_access_key=.*/aws_secret_access_key=${destination_secret_key}/" cloud
```
Comment (ricardomaraschini) [inline scripts/common/object_store.sh:211]: Seems to me that we mixed the "cloud" and the new temp_file variable here, no ? I mean, shouldn't `cloud` be `$temp_file` in the previous two lines ?

Comment (laverya) [reply scripts/common/object_store.sh:211]: yes
yes it should be

Comment (laverya) [inline scripts/common/object_store.sh:211]:
```suggestion
            sed -i "s/aws_access_key_id=.*/aws_access_key_id=${destination_access_key}/" "$temp_file"
            sed -i "s/aws_secret_access_key=.*/aws_secret_access_key=${destination_secret_key}/" "$temp_file"
```

**Review by rrpolanco — APPROVED** [review]

**Pull request merged**